src/apps/halofi/common/halofi.game.constants.ts

```diff
@@ -36,9 +38,69 @@ export const transformRewardArrayToObject = (rewards: RewardBalance[]) => {
   return playerReward;
 };
 
-const ContractVersions = ['2.0.0', '2.0.1', '2.0.2', '2.0.3'];
+/**
+ * queue request and send request with support for throttling
+ * @param {any} request The given request to send
+ * @returns {Promise} A fulfilled promise after the request is processed
+ */
+const queue = new PQueue({ concurrency: 1, interval: 1000 });
+export async function sendRequestWithThrottle(request: any) {
+  return queue.add(() => request);
+}
+
+/**
+ * Wait for the given milliseconds
+ * @param {number} milliseconds The given time to wait
+ * @returns {Promise} A fulfilled promise after the given time has passed
+ */
+function waitFor(milliseconds: number): Promise<any> {
+  return new Promise(resolve => setTimeout(resolve, milliseconds));
+}
+
+/**
+ * Execute a promise and retry with exponential backoff
+ * based on the maximum retry attempts it can perform
+ * @param {function} onRetry callback executed on every retry
+ * @param {number} maxRetries The maximum number of retries to be attempted
+ * @returns {Promise} The result of the given promise passed in
+ */
+export async function retry(onRetry: (...args: any[]) => any, args: any[], maxRetries = 4): Promise<any> {
+  async function retryWithBackoff(retries: number, retryAfter = 1) {
+    try {
+      if (retries > 0) {
+        const timeToWait = 2 ** retries * retryAfter * 1000;
```

Comment:
@meetkosiso 
We should change this part of the code because it will wait a lot more then needed when `retryAfter` is passed w/ the value of `retry-after` header.

consider the following scenario:
1. api request falls into the catch block w/ a 429 error and `retry-after` response header has a value of `14` (seconds)
2. we use this `return retryWithBackoff(retries + 1, retryAfter);` to retry. this passes `retryAfter` which would be 14.
3. Then when calculating `timeToWait` (already using the fixed formula), we'd have the following values:
`timeToWait = (2 ** 2) * 14 * 1000 => 4 * 14 * 1000 = 56 * 1000 => 56000`, so `timetoWait` is `56` seconds, when it should be only `14`.

I think if we get the value from `retry-after` response header, then we could backoff just for that amount of time, rather than calculating the exponential backoff. wdyt?